### PR TITLE
realcugan-ncnn-vulkan: init at 20220728

### PR DESCRIPTION
**Pull request by iynaix**
## Description of changes

ncnn implementation of [Real-CUGAN](https://github.com/bilibili/ailab/blob/main/Real-CUGAN/README_EN.md) converter. Runs fast on Intel / AMD / Nvidia / Apple-Silicon with Vulkan API.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by iynaix** [comment]
Updated to fix build for glslang 14.1.0

**Comment by iynaix** [comment]
Added patch for building on darwin

File: pkgs/by-name/re/realcugan-ncnn-vulkan/package.nix [hunk 1]
```diff
@@ -0,0 +1,67 @@
+{ lib
+, stdenv
```
Comment (FliegendeWurst) [inline pkgs/by-name/re/realcugan-ncnn-vulkan/package.nix:2]: Please format with the new formatter.

File: pkgs/by-name/re/realcugan-ncnn-vulkan/package.nix [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+  installPhase = ''
+    mkdir -p $out/bin $out/share
+    cp realcugan-ncnn-vulkan $out/bin/
+
+    cp -r ${models}/models-{nose,pro,se} $out/share
+  '';
+
+  postFixup = ''
+    patchelf $out/bin/realcugan-ncnn-vulkan --add-needed libvulkan.so
```
Comment (FliegendeWurst) [inline pkgs/by-name/re/realcugan-ncnn-vulkan/package.nix:56]: Needs to be conditional on isLinux

**Review by FliegendeWurst — CHANGES_REQUESTED** [review]

File: pkgs/by-name/re/realcugan-ncnn-vulkan/package.nix (hunk 2 above)
Comment (iynaix) [reply pkgs/by-name/re/realcugan-ncnn-vulkan/package.nix]: I've built and tested the package, no longer seems to need the patchelf to run

File: pkgs/by-name/re/realcugan-ncnn-vulkan/package.nix [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  postPatch = ''
+    substituteInPlace main.cpp --replace REPLACE_MODELS $out/share/models-se
+  '';
+
+  installPhase = ''
+    mkdir -p $out/bin $out/share
+    cp realcugan-ncnn-vulkan $out/bin/
+
+    cp -r ${models}/models-{nose,pro,se} $out/share
+  '';
```
Comment (FliegendeWurst) [inline pkgs/by-name/re/realcugan-ncnn-vulkan/package.nix:60]:
```suggestion
  installPhase = ''
    runHook preInstall
    mkdir -p $out/bin $out/share
    cp realcugan-ncnn-vulkan $out/bin/

    cp -r ${models}/models-{nose,pro,se} $out/share
    runHook postInstall
  '';
```

File: pkgs/by-name/re/realcugan-ncnn-vulkan/package.nix [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    sha256 = "sha256-71C6taL2Zr1exG5HEXOLy1j9ZMKgkMJjTgNi2hiA7xk=";
+  };
+
+  patches = [
+    ./cmakelists.patch
+    ./models_path.patch
+  ];
+
+  cmakeFlags = [
+    "-DUSE_SYSTEM_NCNN=1"
```
Comment (FliegendeWurst) [inline pkgs/by-name/re/realcugan-ncnn-vulkan/package.nix:36]:
```suggestion
    (lib.cmakeBool "USE_SYSTEM_NCNN" true)
```
Similar for others, the last can use lib.cmakeFeature

**Review by FliegendeWurst — CHANGES_REQUESTED** [review]

File: pkgs/by-name/re/realcugan-ncnn-vulkan/package.nix (hunk 4 above)
Comment (iynaix) [reply pkgs/by-name/re/realcugan-ncnn-vulkan/package.nix]: Thanks, made the changes

**Review by FliegendeWurst — APPROVED** [review]

**Pull request merged**